### ISSUE TITLE
Crashing Action on dispose()

### ISSUE THREAD

**Issue by edwardIshaq** [ReactiveCocoa/ReactiveCocoa#2769]
not sure why but after updating Xcode to 7.3 (7D175), Carthage, and building the project again my actions are all crashing 
```

public final class ActionDisposable: Disposable {
	private let action: Atomic<(() -> ())?>

	public var disposed: Bool {
		return action.value == nil
	}

	/// Initializes the disposable to run the given action upon disposal.
	public init(action: () -> ()) {
		self.action = Atomic(action)
	}

	public func dispose() {
		let oldAction = action.swap(nil)
		oldAction?() //<<<<< CRASH with EXEC_BAD_ACCESS
	}
}
```

call site: 
```

    var saveAction: Action<AnyObject?, Event, SaveError> { get }

    @IBAction func next(sender: AnyObject) {
        viewModel
            .saveAction
            .apply(sender)
            .startOnNext{ event in
             self.performSegue()
        }
    }

```

This seems like a memory issue

```
(lldb) bt
* thread #1: tid = 0xaf1b2, 0x00000001116abf4e libdyld.dylib`stack_not_16_byte_aligned_error, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001116abf4e libdyld.dylib`stack_not_16_byte_aligned_error
    frame #1: 0x00007fff547c4b28
    frame #2: 0x000000010ce2ddef ReactiveCocoa`partial apply for thunk [inlined] reabstraction thunk helper from @callee_owned () -> (@unowned ()) to @callee_owned (@in ()) -> (@out ()) + 47 at Signal.swift:0 [opt]
    frame #3: 0x000000010ce2dde9 ReactiveCocoa`partial apply for thunk + 41 at Signal.swift:0 [opt]
    frame #4: 0x000000010ce2d7d0 ReactiveCocoa`protocol witness for Disposable.dispose() -> () in conformance ActionDisposable + 80 at Disposable.swift:49 [opt]
    frame #5: 0x000000010ce2d7c5 ReactiveCocoa`protocol witness for Disposable.dispose() -> () in conformance ActionDisposable + 69 at Signal.swift:47 [opt]
    frame #6: 0x000000010ce7710d ReactiveCocoa`SignalProducer.(event=<unavailable>, producerDisposable=0x00007ffe93253e60) -> ()) -> ()).(closure #2) + 301 at SignalProducer.swift:257 [opt]
    frame #7: 0x000000010ce6a06e ReactiveCocoa`partial apply for SignalProducer.(startWithSignal((Signal<A, B>, Disposable) -> ()) -> ()).(closure #2) + 110 at SignalProducer.swift:0 [opt]
    frame #8: 0x000000010ce3d1a0 ReactiveCocoa`Signal.(event=<unavailable>, interrupted=(spinLock = 0, _value = false), sendLock=0x00007ffe932612e0, self=0x00007ffe93261da0, generatorDisposable=0x00007ffe93261380) -> Disposable?) -> Signal<A, B>).(closure #1) + 2000 at Signal.swift:63 [opt]
    frame #9: 0x000000010ce438ee ReactiveCocoa`SignalType.(receivedEvent=<unavailable>, event=nil, next=0x000000010c711070 PopFramework`partial apply forwarder for reabstraction thunk helper <A, B where B: Swift.ErrorType> from @callee_owned (@in A) -> (@unowned ()) to @callee_owned (@in A) -> (@out ()) at ReactiveCocoaExtensions.swift, failed=nil, completed=0x0000000000020000, interrupted=0x000000011d211000, terminated=nil) -> ()?, failed : (A.Error) -> ()?, completed : () -> ()?, interrupted : () -> ()?, terminated : () -> ()?, disposed : () -> ()?, next : (A.Value) -> ()?) -> Signal<A.Value, A.Error>).(closure #1).(closure #1) + 1150 at Signal.swift:537 [opt]
    frame #10: 0x000000010ce39db8 ReactiveCocoa`partial apply for SignalType.(on(event : (Event<A.Value, A.Error>) -> ()?, failed : (A.Error) -> ()?, completed : () -> ()?, interrupted : () -> ()?, terminated : () -> ()?, disposed : () -> ()?, next : (A.Value) -> ()?) -> Signal<A.Value, A.Error>).(closure #1).(closure #1) + 440 at Signal.swift:0 [opt]
    frame #11: 0x000000010ce3d1a0 ReactiveCocoa`Signal.(event=<unavailable>, interrupted=(spinLock = 0, _value = false), sendLock=0x00007ffe93261b50, self=0x00007ffe93260f50, generatorDisposable=0x00007ffe93260fb0) -> Disposable?) -> Signal<A, B>).(closure #1) + 2000 at Signal.swift:63 [opt]
    frame #12: 0x000000010ce770bd ReactiveCocoa`SignalProducer.(event=<unavailable>, producerDisposable=0x00007ffe93261c70) -> ()) -> ()).(closure #2) + 221 at SignalProducer.swift:252 [opt]
    frame #13: 0x000000010ce6a06e ReactiveCocoa`partial apply for SignalProducer.(startWithSignal((Signal<A, B>, Disposable) -> ()) -> ()).(closure #2) + 110 at SignalProducer.swift:0 [opt]
    frame #14: 0x000000010ce8b8e5 ReactiveCocoa`Observer.sendFailed(error=0x00007ffe93261830) -> () + 149 at Observer.swift:45 [opt]
    frame #15: 0x000000010cebeb89 ReactiveCocoa`Action.(disposable=0x00007ffe93261c70, self=0x00007ffe930874e0, input=0x00007ffe93254df0) -> SignalProducer<B, ActionError<C>>).(closure #1) + 1369 at Action.swift:112 [opt]
    frame #16: 0x000000010ce5f957 ReactiveCocoa`SignalProducer.startWithSignal((Signal<A, B>, Disposable) -> ()) -> () + 759 at SignalProducer.swift:261 [opt]
    frame #17: 0x000000010ce31397 ReactiveCocoa`protocol witness for SignalProducerType.startWithSignal((Signal<A.Value, A.Error>, Disposable) -> ()) -> () in conformance <A, B where ...> SignalProducer<A, B> + 87 at SignalProducer.swift:232 [opt]
    frame #18: 0x000000010ce7f5ea ReactiveCocoa`SignalProducerType.(compositeDisposable=0x00007ffe93253e60, started=nil, self=0x00007ffe932630f0 -> 0x000000010ceb9f30 ReactiveCocoa`partial apply forwarder for ReactiveCocoa.Action.(apply (A) -> ReactiveCocoa.SignalProducer<B, ReactiveCocoa.ActionError<C>>).(closure #1) at Action.swift, event=nil, failed=nil, completed=nil, interrupted=nil, terminated=nil, disposed=nil, next=0x000000010c711070 PopFramework`partial apply forwarder for reabstraction thunk helper <A, B where B: Swift.ErrorType> from @callee_owned (@in A) -> (@unowned ()) to @callee_owned (@in A) -> (@out ()) at ReactiveCocoaExtensions.swift) -> ()?, event : (Event<A.Value, A.Error>) -> ()?, failed : (A.Error) -> ()?, completed : () -> ()?, interrupted : () -> ()?, terminated : () -> ()?, disposed : () -> ()?, next : (A.Value) -> ()?) -> SignalProducer<A.Value, A.Error>).(closure #1) + 682 at SignalProducer.swift:826 [opt]
    frame #19: 0x000000010ce6bef2 ReactiveCocoa`partial apply for SignalProducerType.(on(started : () -> ()?, event : (Event<A.Value, A.Error>) -> ()?, failed : (A.Error) -> ()?, completed : () -> ()?, interrupted : () -> ()?, terminated : () -> ()?, disposed : () -> ()?, next : (A.Value) -> ()?) -> SignalProducer<A.Value, A.Error>).(closure #1) + 610 at SignalProducer.swift:0 [opt]
    frame #20: 0x000000010ce5f957 ReactiveCocoa`SignalProducer.startWithSignal((Signal<A, B>, Disposable) -> ()) -> () + 759 at SignalProducer.swift:261 [opt]
    frame #21: 0x000000010ce31397 ReactiveCocoa`protocol witness for SignalProducerType.startWithSignal((Signal<A.Value, A.Error>, Disposable) -> ()) -> () in conformance <A, B where ...> SignalProducer<A, B> + 87 at SignalProducer.swift:232 [opt]
    frame #22: 0x000000010ce5fa69 ReactiveCocoa`SignalProducerType.start(self=0x00007fff547c5950 -> 0x000000010ce6bc90 ReactiveCocoa`partial apply forwarder for (extension in ReactiveCocoa):ReactiveCocoa.SignalProducerType.(on (started : Swift.Optional<() -> ()>, event : Swift.Optional<(ReactiveCocoa.Event<A.Value, A.Error>) -> ()>, failed : Swift.Optional<(A.Error) -> ()>, completed : Swift.Optional<() -> ()>, interrupted : Swift.Optional<() -> ()>, terminated : Swift.Optional<() -> ()>, disposed : Swift.Optional<() -> ()>, next : Swift.Optional<(A.Value) -> ()>) -> ReactiveCocoa.SignalProducer<A.Value, A.Error>).(closure #1) at SignalProducer.swift) -> Disposable + 153 at SignalProducer.swift:323 [opt]
    frame #23: 0x000000010c711b31 PopFramework`SignalProducer.startOnNext(next=0x000000010b560e30 Pop`partial apply forwarder for reabstraction thunk helper from @callee_owned (@owned Pop.Event) -> (@unowned ()) to @callee_owned (@in Pop.Event) -> (@unowned ()) at EventSelectTypeVC.swift, self=(startHandler = 0x000000010ceb9f30 ReactiveCocoa`partial apply forwarder for ReactiveCocoa.Action.(apply (A) -> ReactiveCocoa.SignalProducer<B, ReactiveCocoa.ActionError<C>>).(closure #1) at Action.swift)) -> ()) -> Disposable + 257 at ReactiveCocoaExtensions.swift:82
  * frame #24: 0x000000010b560d11 Pop`EventSelectTypeVC.next(sender=0x00007ffe931b69e0, self=0x00007ffe931b6b70) -> () + 449 at EventSelectTypeVC.swift:92
    frame #25: 0x000000010b560eb6 Pop`@objc EventSelectTypeVC.next(AnyObject) -> () + 54 at EventSelectTypeVC.swift:0
    frame #26: 0x000000010fa23a8d UIKit`-[UIApplication sendAction:to:from:forEvent:] + 92
    frame #27: 0x000000010fe31067 UIKit`-[UIBarButtonItem(UIInternal) _sendAction:withEvent:] + 152
    frame #28: 0x000000010fa23a8d UIKit`-[UIApplication sendAction:to:from:forEvent:] + 92
    frame #29: 0x000000010fb96e67 UIKit`-[UIControl sendAction:to:forEvent:] + 67
    frame #30: 0x000000010fb97143 UIKit`-[UIControl _sendActionsForEvents:withEvent:] + 327
    frame #31: 0x000000010fb972be UIKit`-[UIControl _sendActionsForEvents:withEvent:] + 706
    frame #32: 0x000000010fb96263 UIKit`-[UIControl touchesEnded:withEvent:] + 601
    frame #33: 0x000000010fa9699f UIKit`-[UIWindow _sendTouchesForEvent:] + 835
    frame #34: 0x000000010fa976d4 UIKit`-[UIWindow sendEvent:] + 865
    frame #35: 0x000000010fa42dc6 UIKit`-[UIApplication sendEvent:] + 263
    frame #36: 0x000000012085fb15 UIKit`-[UIApplicationAccessibility sendEvent:] + 77
    frame #37: 0x000000010fa1c553 UIKit`_UIApplicationHandleEventQueue + 6660
    frame #38: 0x000000010e364301 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
    frame #39: 0x000000010e35a22c CoreFoundation`__CFRunLoopDoSources0 + 556
    frame #40: 0x000000010e3596e3 CoreFoundation`__CFRunLoopRun + 867
    frame #41: 0x000000010e3590f8 CoreFoundation`CFRunLoopRunSpecific + 488
    frame #42: 0x0000000113ac2ad2 GraphicsServices`GSEventRunModal + 161
    frame #43: 0x000000010fa21f09 UIKit`UIApplicationMain + 171
    frame #44: 0x000000010b643422 Pop`main + 114 at AppDelegate.swift:21
    frame #45: 0x00000001116ad92d libdyld.dylib`start + 1
    frame #46: 0x00000001116ad92d libdyld.dylib`start + 1
(lldb) 
```

**Comment by edwardIshaq** [comment]
my Cartfile:

$ cat Cartfile
github "ReactiveCocoa/ReactiveCocoa"
github "SwiftyJSON/SwiftyJSON"
github "CocoaLumberjack/CocoaLumberjack"
github "realm/realm-cocoa"

$ cat Cartfile.resolved 
github "CocoaLumberjack/CocoaLumberjack" "2.2.0"
github "antitypical/Result" "1.0.2"
github "SwiftyJSON/SwiftyJSON" "2.3.3"
github "realm/realm-cocoa" "v0.98.6"
github "ReactiveCocoa/ReactiveCocoa" "v4.0.1"

**Comment by edwardIshaq** [comment]
I've created a simple Gist that captures the problem:
https://gist.github.com/edwardIshaq/7ada399d8784416176d5

from stepping into the code I'm seeing the following behavior :
```

	public func apply(input: Input) -> SignalProducer<Output, ActionError<Error>> {
		return SignalProducer { observer, disposable in
			var startedExecuting = false

			dispatch_sync(self.executingQueue) {
				if self._enabled.value {
					self._executing.value = true
					startedExecuting = true
				}
			}

			if !startedExecuting {
				observer.sendFailed(.NotEnabled)
				return
			}

			self.executeClosure(input).startWithSignal { signal, signalDisposable in
				disposable.addDisposable(signalDisposable)

				signal.observe { event in
					observer.action(event.mapError { .ProducerError($0) })
					self.eventsObserver.sendNext(event)
				}
			}

			disposable.addDisposable {
				self._executing.value = false
			}
		}
	}

```

when apply is called on the action the code is calling: 
```
			if !startedExecuting {
				observer.sendFailed(.NotEnabled)
				return
			}

```

before it had a chance to execute on the executingQueue so its failing and disposing immediately.
Not sure if this is by design but I'm completely block by this crash.

**Comment by mdiep** [comment]
You're most likely hitting #2751.

tl;dr: Because of a bug in Swift 2.2, the current RAC release has a bug. There's a fix on `master`, but a new release is waiting on a new release of Result.

**Comment by edwardIshaq** [comment]
thats what i ended up doing, After compiling with Master it worked fine.
